src/getINP.ts

```diff
@@ -14,118 +14,127 @@
-    // ...then the inpEntry will change, but `duration` value of the new entry may still be the same.
-    // While technically the INP metric.value doesn't change, we still report since metric.entries changes.
-    //
-    // Potentially, we may even want to compare the whole metric.entries range for equality, because:
-    // * We can have cases where a middle value updates due to new-longest value with same interactionId.
-    // * When we are already at MAX_ENTRIES and `which` stops changing, but the current smallest can get popped off.
-    const which = largestINPEntries.indexOf(inpEntry);
-    if (which >= metric.entries.length || metric.value != inpEntry.duration) {
-      metric.value = inpEntry.duration;
-      // We attach all the longest responsiveness entries, not just the HighP value.
-      // While technically the INP score is exactly the entry.duration of one specific HighP-ile entry...
-      // the entry would not have been picked (and IMP would be lower) if *any* of the worst entries were not so high.
-      // Improving any of them will improve score.
-      metric.entries.length = 0;
-      metric.entries.push(...largestINPEntries.slice(0, which + 1));
-    }
+  const handleEntries = (entries: Metric['entries']) => {
+    (entries as PerformanceEventTiming[]).forEach((entry) => {
+      if (entry.interactionId) {
+        processEntry(entry);
+      }
+    });
 
-    // Perhaps Event Timing is the first API that can have multiple entries in a single PO callback
-    // That means that we would ideally report() only after the whole list of entries is processed, not one per entry.
-    // If we were lucky, the entries would be in timestamp order so the first is the longest... but I've found they
-    // are ordered in other ways... by type, I think?
-    // Alternatively: sort entries in the observe() wrapper.
-    report();
+    const inp = estimateP98LongestInteraction();
+
+    if (inp && inp.latency !== metric.value) {
+      metric.value = inp.latency;
+      metric.entries = inp.entries;
+      report();
+    }
   };
 
-  const po = observe('event', entryHandler as PerformanceEntryHandler);
+  const po = observe('event', handleEntries, {
+    // The use of 50 here as a balance between not wanting the callback to
+    // run for too many already fast events (one frame or less), but also
+    // get enough fidelity below the recommended "good" threshold of 200.
+    durationThreshold: 50,
+  } as PerformanceObserverInit);
+
   report = bindReporter(onReport, metric, reportAllChanges);
 
   if (po) {
     onHidden(() => {
-      po.takeRecords().map(entryHandler as PerformanceEntryHandler);
+      handleEntries(po.takeRecords());
+
+      // If the interaction count shows that there were interactions but
+      // none were captured by the PerformanceObserver, report a latency of 0.
+      if (metric.value < 0 &&
+          getInteractionCount() - prevInteractionCount > 0) {
```

Comment:
This seems potentially brittle.  Could you create a simple helper called getInteractionCountForCurrentNavigation() or something, or is this specifically inlined to reduce code size?

Reply:
I can add that. I agree it would be less brittle.

Reply:
Done.